File: crates/klyron_ai/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::AiBuilder::Impl {
public:
  Klyron::AiConfig config;
  bool verbose = false;
};

Klyron::AiBuilder::Klyron::AiBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::AiBuilder& Klyron::AiBuilder::withConfig(const Klyron::AiConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::AiBuilder& Klyron::AiBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::AiBuilder::Klyron::AiInstance Klyron::AiBuilder::build() {
  Klyron::AiInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_ai/bindings/rust/builder.rs
//! Builder pattern for klyron_ai

use crate::types::Klyron::AiConfig;

#[derive(Debug, Default)]
pub struct Klyron::AiBuilder {
    config: Option<Klyron::AiConfig>,
    verbose: bool,
}

impl Klyron::AiBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::AiConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::AiInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::AiInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::AiInstance {
    pub config: Klyron::AiConfig,
    pub verbose: bool,
}
